### PR TITLE
#4890 feat: replace usage of flowcrypt.com/api/ with flowcrypt.com/shared-tenant-fes/

### LINKED ISSUE #4890: replace usage of flowcrypt.com/api/ with flowcrypt.com/shared-tenant-fes/
reasoning here https://github.com/FlowCrypt/flowcrypt-enterprise-on-prem/issues/4097

On this repository, we have two account servers: `EnterpriseServer` (at fes.domain.com) and `FlowCryptComApi` (at flowcrypt.com/api). I have just deployed https://flowcrypt.com/shared-tenant-fes/ (same app as the already implemented `EnterpriseServer` but different instance - as a replacement of flowcrypt.com/api)

In this issue:
 - [x] rename `EnterpriseServer` to `ExternalService` (and `FlowCrypt Enterprise Server` to `FlowCrypt External Service` and, `enterprise-server.ts` to `external-service.ts` so on)
 - [x] in this `ExternalService` top comment change `fully independent of flowcrypt.com/api` to `fully independent of flowcrypt.com/shared-tenant-fes/ ` and `FlowCrypt Enterprise Server (FES) may be deployed on-prem by enterprise customers.` to `FlowCrypt External Service (FES) can be run individually by each customer at fes.customer.com. When not present, flowcrypt.com/shared-tenant-fes/ will be used.`
 - [x] remove all usage of `FlowCryptComApi.accountUpdate`. Where there is UI that uses this, remove the UI and its handlers (eg changing message expiration time).
 - [x] delete `FlowCryptComApi` class
 - [x] helpers.ts/isFesUsed should be renamed to `isCustomUrlFesUsed` along with usage
 - [x] in constructor of `AccountServer` add three lines: 
```ts
this.potentialCustomUrlFes = new ExternalService(this.acctEmail);`
this.sharedTenantFes = new ExternalService(this.acctEmail);
this.sharedTenantFes.url = 'https://flowcrypt.com/shared-tenant-fes/';
```
 - [x] remove all `BackendRes` types. Where these types are still used, like `AccountServer.accountGetAndUpdateLocalStore`, create a type instead. This one in particular can return `{clientConfiguration, defaultWebPortalMessageExpire: number}`
 - [x] `accountGetAndUpdateLocalStore` should continue returning `180` days when `isCustomUrlFesUsed` else it should return `90` when using `sharedTenantUrl`
 - [x] `AccountServer.accountUpdate` should be removed
 - [x] `AccountUpdate` methods `messageUpload`, `messageGatewayUpdate`, `messageToken` should all have `else` branches that look similar as the `if` branch, except using `sharedTenantFes` to do the same thing
 - I suspect the `AccountUpdate` class will look odd this way, with a lot of repetition, but I propose to not refactor that further yet. It will make review of this PR much easier, to ensure a mistake wasn't made (it's crucial functionality, and can be easily subtly wrong). Then we can have a separate PR that will refactor `AccountUpdate` further.
 - [x] tests themselves SHOULD NOT need to be udpated (I hope), except that mock definitions need to be fixed. In particular, all endpoints on `mockBackendEndpoints` except for `/api/help/error` and `/api/help/feedback` can be removed, and:
 - [x] move file `fes-endpoints.ts` to `custom-url-fes-endpoints.ts` and likewise rename the exported variable
 - [x]  create file shared-tenant-fes-endpoints.ts that is similar to custom-url-fes-endpoints.ts (starting from line export const mockFesEndpoints: HandlersDefinition = { downwards can be copied and edited) except there will be none of the special cases. Name it mockSharedTenantFesEndpoints. Also need to replace in this new file /api/ with /shared-tenant-fes/api/. Otherwise just return one default response per endpoints, always the same, except for client configuraiton. Still need to keep returning request-based client configuration. Previously backend-endpoints was using mockBackendData.getClientConfiguration(email!). You can delete mockBackendData completely, and instead copy contents of mockBackendData.getClientConfiguration into the newly created custom-url-fes-endpoints.ts / /shared-tenant-fes/api/v1/client-configuration endpoint, editing it to ensure it will return correct format.
 - [x] in `all-apis-mock.ts` the `mockFesEndpoints` should be renamed to `mockCustomUrlFesEndpoints`, and underneath should add `mockSharedTenantFesEndpoints`

I've tried to list all of the steps to take to make implementation easier, it is quite an involved issue.

### PR DESCRIPTION
**Pull request by ioanmo226**
This PR replaced usage of flowcrypt.com/api/ with flowcrypt.com/shared-tenant-fes/

close #4890 // if this PR closes an issue

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities
- [ ] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

File: extension/chrome/settings/index.ts [hunk 1]
```diff
@@ -370,7 +370,8 @@ View.run(
           // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
           const acctRes = await this.acctServer!.accountGetAndUpdateLocalStore();
           $('#status-row #status_flowcrypt').text(`fc:ok`);
-          if (acctRes?.account?.alias) {
+          // if (acctRes?.account?.alias) { // Todo: get alias
+          if (acctRes) {
             statusContainer.find('.status-indicator-text').css('display', 'none');
             statusContainer.find('.status-indicator').addClass('active');
           } else {
```
Comment (ioanmo226) [inline extension/chrome/settings/index.ts:377]: @tomholub Now accountGetAndUpdateLocalStore only returns clientConfigruation and defaultWebPortalMessageExpire.
How this should be handled?
I think this status-indicator is for encrypted contact page?

Comment (tomholub) [reply extension/chrome/settings/index.ts:377]: Let's remove the status indicator

Comment (ioanmo226) [reply extension/chrome/settings/index.ts:377]: ok

File: extension/js/common/api/account-server.ts [hunk 2]
```diff
@@ -2,55 +2,46 @@
...
-      await fes.accountUpdate(profileUpdate);
-    } else {
-      await FlowCryptComApi.accountUpdate(await this.getIdToken(), profileUpdate);
-    }
+  public accountGetAndUpdateLocalStore = async (): Promise<AccountGetAndUpdateResult> => {
+    const service = await this.getExternalService();
+    const fetchedClientConfiguration = await service.fetchAndSaveClientConfiguration();
+    return {
+      clientConfiguration: fetchedClientConfiguration,
+      defaultWebPortalMessageExpire: (await this.isFesUsed()) ? 180 : 90,
```
Comment (tomholub) [inline extension/js/common/api/account-server.ts:43]: Looking at this now, `defaultWebPortalMessageExpire` no longer has to be here, either. I'd remove it from here, and instead move it to where it's actually used. If it's more than one place where it's used, then there should be a method somewhere called `getWebPortalMessageExpireDays` which will either return 180 or 90. The method could possibly be on this class, for now. 

Also `AccountGetAndUpdateResult` could be removed, this method can be called `fetchAndSaveClientConfiguration`  as well, instead of `accountGetAndUpdateLocalStore`, and it can return `ClientConfigurationJson` directly.

File: extension/js/common/api/account-server.ts [hunk 3]
```diff
@@ -60,44 +51,28 @@ export class AccountServer extends Api {
...
-    if (!idToken) {
-      // user will not actually see this message, they'll see a generic login prompt
-      throw new BackendAuthErr('Missing id token, please re-authenticate');
+  private getExternalService = async (): Promise<ExternalService> => {
+    if (await this.isFesUsed()) {
+      return this.potentialCustomUrlFes;
     }
-    return idToken;
+    return this.sharedTenantFes;
   };
```
Comment (tomholub) [inline extension/js/common/api/account-server.ts:77]: This looks good to me.

File: extension/js/common/api/account-server.ts [hunk 4]
```diff
@@ -60,44 +51,28 @@ export class AccountServer extends Api {
...
-      return { replyToken: res.token };
-    }
+    const service = await this.getExternalService();
+    return await service.webPortalMessageNewReplyToken();
   };
 
   public isFesUsed = async (): Promise<boolean> => {
-    return await isFesUsed(this.acctEmail);
+    return await isCustomUrlFesUsed(this.acctEmail);
   };
```
Comment (tomholub) [inline extension/js/common/api/account-server.ts:70]: If you want to keep this method here (there's not much in it, could possibly be removed) then it should be also called `isCustomUrlFesUsed` instead of  `isFesUsed`

File: extension/js/common/api/account-servers/external-service.ts [hunk 5]
```diff
@@ -29,12 +28,12 @@ export namespace FesRes {
 }
 
 /**
- * FlowCrypt Enterprise Server (FES) may be deployed on-prem by enterprise customers.
+ * FlowCrypt External Service (FES) may be deployed on-prem by enterprise customers.
  * This gives them more control. All OrgRules, log collectors, etc (as implemented) would then be handled by the FES.
```
Comment (tomholub) [inline extension/js/common/api/account-servers/external-service.ts:32]: `OrgRules` -> `Client Configuration`

File: extension/js/common/api/account-servers/external-service.ts [hunk 6]
```diff
@@ -29,12 +28,12 @@ export namespace FesRes {
...
  * Once fully integrated, this will allow customers to be fully independent of flowcrypt.com/api
```
Comment (tomholub) [inline extension/js/common/api/account-servers/external-service.ts:33]: `of flowcrypt.com/shared-tenant-fes`

File: extension/js/common/helpers.ts [hunk 7]
```diff
@@ -13,7 +13,7 @@ import { PassphraseStore } from './platform/store/passphrase-store.js';
 import { Bm } from './browser/browser-msg.js';
 import { PgpPwd } from './core/crypto/pgp/pgp-password.js';
 
-export const isFesUsed = async (acctEmail: string) => {
+export const isCustomUrlFesUsed = async (acctEmail: string) => {
```
Comment (tomholub) [inline extension/js/common/helpers.ts:16]: Also, let's change everywhere that is currently `CustomUrlFes` to become `CustomerUrlFes` I think that makes the name clearer (the customer runs that FES themselves)

**Review by tomholub — COMMENTED** [review]
partial review, looking good

File: extension/js/common/api/account-server.ts (hunk 4 above)
Comment (ioanmo226) [reply extension/js/common/api/account-server.ts:70]: ok

File: extension/js/common/helpers.ts (hunk 7 above)
Comment (ioanmo226) [reply extension/js/common/helpers.ts:16]: ok

File: extension/js/common/api/account-server.ts (hunk 2 above)
Comment (ioanmo226) [reply extension/js/common/api/account-server.ts]: Sounds good.

File: extension/chrome/elements/compose-modules/formatters/encrypted-mail-msg-formatter.ts [hunk 8]
```diff
@@ -175,17 +174,7 @@ export class EncryptedMsgMailFormatter extends BaseMailFormatter {
...
-     * Another thing to note is that eventually, flowcrypt.com/api web portal functionality will
-     *   be deprecated, and we'll instead run a "shared tenant FES instance" to fill that role.
-     *   Nothing will change for users, but our code on the client will be more streamlined.
-     *   Therefore, eventually, this `if` branch with the line below will be removed once both
-     *   consumers and enterprises use API with the same structure.
      */
-    if (!(await this.view.acctServer.isFesUsed())) {
-      // if flowcrypt.com/api is used
-      newMsg.pwd = await PgpHash.challengeAnswer(newMsg.pwd); // then hash the password to preserve compatibility
-    }
```
Comment (ioanmo226) [inline extension/chrome/elements/compose-modules/formatters/encrypted-mail-msg-formatter.ts:188]: @tomholub I also removed this `if` branch as per (removed) above comment.
Am I correct?

Comment (tomholub) [reply extension/chrome/elements/compose-modules/formatters/encrypted-mail-msg-formatter.ts:188]: you are correct here

**Comment by ioanmo226** [comment]
Ready for review.
Please check carefully as I might have missed something.

File: extension/chrome/elements/compose-modules/compose-err-module.ts [hunk 9]
```diff
@@ -157,8 +157,8 @@ export class ComposeErrModule extends ViewModule<ComposeView> {
   };
 
   public throwIfEncryptionPasswordInvalidOrDisabled = async ({ subject, pwd }: { subject: string; pwd?: string }) => {
-    // When DISABLE_FLOWCRYPT_HOSTED_PASSWORD_MESSAGES present, and recipients are missing a public key, and the user is using flowcrypt.com/api (not FES)
-    if (this.view.clientConfiguration.shouldDisablePasswordMessages() && !this.view.isFesUsed()) {
+    // When DISABLE_FLOWCRYPT_HOSTED_PASSWORD_MESSAGES present, and recipients are missing a public key, and the user is using flowcrypt.com/shared-tenant-fes (not FES)
+    if (this.view.clientConfiguration.shouldDisablePasswordMessages() && !this.view.isCustomUrlFesUsed()) {
```
Comment (tomholub) [inline extension/chrome/elements/compose-modules/compose-err-module.ts:161]: Should rename to `shouldDisableFlowCryptHostedPasswordMessages` to reflect hos it's used

File: extension/chrome/elements/compose-modules/compose-pwd-or-pubkey-container-module.ts [hunk 10]
```diff
@@ -139,7 +140,7 @@ export class ComposePwdOrPubkeyContainerModule extends ViewModule<ComposeView> {
   };
 
   private showMsgPwdUiAndColorBtn = async (anyNopgp: boolean, anyRevoked: boolean) => {
-    const isPasswordMessageDisabled = this.view.clientConfiguration.shouldDisablePasswordMessages() && !this.view.isFesUsed();
+    const isPasswordMessageDisabled = this.view.clientConfiguration.shouldDisablePasswordMessages() && !this.view.isCustomUrlFesUsed();
```
Comment (tomholub) [inline extension/chrome/elements/compose-modules/compose-pwd-or-pubkey-container-module.ts:143]: Let's call it `isCustomerUrlFesUsed`

File: extension/chrome/elements/compose-modules/formatters/encrypted-mail-msg-formatter.ts [hunk 11]
```diff
@@ -94,8 +93,8 @@ export class EncryptedMsgMailFormatter extends BaseMailFormatter {
 
   private formatSendablePwdMsgs = async (newMsg: NewMsgData, pubkeys: PubkeyResult[], signingKey?: ParsedKeyInfo) => {
     // password-protected message, temporarily uploaded (already encrypted) to:
-    //    - flowcrypt.com/api (consumers and customers without on-prem setup), or
-    //    - FlowCrypt Enterprise Server (enterprise customers with on-prem setup)
+    //    - flowcrypt.com/shared-tenant-fes (consumers and customers without on-prem setup), or
+    //    - FlowCrypt External Service (enterprise customers with on-prem setup)
```
Comment (tomholub) [inline extension/chrome/elements/compose-modules/formatters/encrypted-mail-msg-formatter.ts:97]: `FlowCrypt External Service` here can be replaced with `fes.customer-domain.com`

File: extension/js/common/api/account-servers/external-service.ts [hunk 12]
```diff
@@ -157,11 +156,6 @@ export class EnterpriseServer extends Api {
     });
   };
 
-  public accountUpdate = async (profileUpdate: ProfileUpdate): Promise<BackendRes.FcAccountUpdate> => {
-    console.log('profile update ignored', profileUpdate);
-    throw new UnreportableError('Account update not implemented when using FlowCrypt Enterprise Server');
-  };
-
   private authHdr = async (): Promise<Dict<string>> => {
```
Comment (tomholub) [inline extension/js/common/api/account-servers/external-service.ts:159]: In another issue, I'd like all calls to FES to include the `authHdr`.

File: extension/js/common/api/account-servers/external-service.ts [hunk 13]
```diff
@@ -29,12 +28,12 @@ export namespace FesRes {
 }
 
 /**
- * FlowCrypt Enterprise Server (FES) may be deployed on-prem by enterprise customers.
- * This gives them more control. All OrgRules, log collectors, etc (as implemented) would then be handled by the FES.
- * Once fully integrated, this will allow customers to be fully independent of flowcrypt.com/api
+ * FlowCrypt External Service (FES) may be deployed on-prem by enterprise customers.
+ * This gives them more control. All Client Configurations, log collectors, etc (as implemented) would then be handled by the FES.
+ * Once fully integrated, this will allow customers to be fully independent of flowcrypt.com/shared-tenant-fes
```
Comment (tomholub) [inline extension/js/common/api/account-servers/external-service.ts:33]: Please change `Once fully integrated, this will allow` to `This allows`

File: extension/js/common/api/email-provider/gmail/google-auth.ts [hunk 14]
```diff
@@ -176,25 +176,25 @@ export class GoogleAuth {
...
+          await acctServer.fetchAndSaveClientConfiguration();
         } else {
           // eventually this branch will be dropped once a public FES instance is run for these customers
-          // when using flowcrypt.com/api, pulling ClientConfiguration is authenticated, therefore have
+          // when using flowcrypt.com/shared-tenant-fes, pulling ClientConfiguration is authenticated, therefore have
           //   to retrieve access token first (which is the only way to authenticate other calls)
           const acctServer = new AccountServer(authRes.acctEmail);
           // fetch and store ClientConfiguration (authenticated)
-          await acctServer.accountGetAndUpdateLocalStore(); // stores ClientConfiguration
+          await acctServer.fetchAndSaveClientConfiguration(); // stores ClientConfiguration
```
Comment (tomholub) [inline extension/js/common/api/email-provider/gmail/google-auth.ts:193]: There is an opportunity to simplify here it seems, I think the two branches don't quite have to be two branches anymore, but unfortunately I can't provide exact details on what to do here without sitting down and trying to rewrite it myself. Please try taking a look.

Comment (tomholub) [reply extension/js/common/api/email-provider/gmail/google-auth.ts:193]: We could alternatively do it in another PR.. for safety.

File: extension/js/common/lang.ts [hunk 15]
```diff
@@ -6,10 +6,11 @@ import { FLAVOR } from './core/const.js';
 
 const isEnterpriseBuildUsed = FLAVOR === 'enterprise';
 
-const contactMinimalSubsentence = (isFesUsed: boolean) => (isFesUsed || isEnterpriseBuildUsed ? 'contact your Help Desk' : 'write us at human@flowcrypt.com');
-const contactIfHappensAgain = (isFesUsed: boolean) => `If this happens again, please ${contactMinimalSubsentence(isFesUsed)}. `;
-const contactForSupportSubsentence = (isFesUsed: boolean, reason = '') =>
-  isFesUsed || isEnterpriseBuildUsed ? `Please contact your Help Desk ${reason}` : `Please write us at human@flowcrypt.com ${reason}`;
+const contactMinimalSubsentence = (isCustomUrlFesUsed: boolean) =>
```
Comment (tomholub) [inline extension/js/common/lang.ts:9]: `isCustomerUrlFesUsed` plus many more times below.

File: test/source/mock/backend/backend-data.ts [hunk 16]
```diff
@@ -38,35 +38,19 @@ export interface ReportedError {
 export class BackendData {
   public reportedErrors: ReportedError[] = [];
 
-  public clientConfigurationByAcctEmail: Dict<ClientConfiguration | HttpClientErr> = {};
+  public forceClientConfigurationForDomain: Dict<ClientConfiguration | HttpClientErr> = {};
```
Comment (tomholub) [inline test/source/mock/backend/backend-data.ts:41]: Not too sure about the `force` word, I don't see anything being forced

File: test/source/mock/fes/custom-url-fes-endpoints.ts [hunk 17]
```diff
@@ -185,7 +185,7 @@ const processMessageFromUser4 = async (body: string) => {
   return response;
```
Comment (tomholub) [inline test/source/mock/fes/custom-url-fes-endpoints.ts:185]: name of file `customer` instead of `custom`

File: test/source/mock/fes/shared-tenant-fes-endpoints.ts [hunk 18]
```diff
@@ -0,0 +1,128 @@
+/* ©️ 2016 - present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com */
+
+import { expect } from 'chai';
+import { IncomingMessage } from 'http';
+import { HandlersDefinition } from '../all-apis-mock';
+import { HttpClientErr, Status } from '../lib/api';
+import { MockJwt } from '../lib/oauth';
+import { mockBackendData } from '../backend/backend-endpoints';
+
+const standardFesUrl = 'fes.standardsubdomainfes.localhost:8001';
```
Comment (tomholub) [inline test/source/mock/fes/shared-tenant-fes-endpoints.ts:10]: This is not relevant in shared-tenant-fes context. That's url of a customer deployed fes.

File: test/source/mock/google/strategies/send-message-strategy.ts [hunk 19]
```diff
@@ -38,15 +38,15 @@ class PwdAndPubkeyEncryptedMessagesWithFlowCryptComApiTestStrategy implements IT
...
       expect(mimeMsg.bcc).to.be.an.undefined;
       expect(mimeMsg.to).to.be.an.undefined;
       expect((mimeMsg.headers.get('reply-to') as AddressObject).text).to.equal('First Last <flowcrypt.compatibility@gmail.com>, test@email.com');
     } else {
       expect(mimeMsg.text!).to.contain(`${senderEmail} has sent you a password-encrypted email`);
       expect(mimeMsg.text!).to.contain('Follow this link to open it');
       if (!mimeMsg.text?.match(/https:\/\/flowcrypt.com\/[a-z0-9A-Z]{10}/)) {
-        throw new HttpClientErr(`Error: cannot find pwd encrypted flowcrypt.com/api link in:\n\n${mimeMsg.text}`);
+        throw new HttpClientErr(`Error: cannot find pwd encrypted flowcrypt.com/shared-tenant-fes link in:\n\n${mimeMsg.text}`);
       }
```
Comment (tomholub) [inline test/source/mock/google/strategies/send-message-strategy.ts:50]: The link is expected to be different now, definitely not `/https:\/\/flowcrypt.com\/[a-z0-9A-Z]{10}/`. Try sending a password protected message. It will be something like `https://flowcrypt.com/shared-tentant-fes/message/...`

File: test/source/mock/google/strategies/send-message-strategy.ts [hunk 20]
```diff
@@ -59,7 +59,7 @@ class PwdEncryptedMessageWithFlowCryptComApiTestStrategy implements ITestMsgStra
       throw new HttpClientErr(`Error checking sent text in:\n\n${mimeMsg.text}`);
     }
     if (!mimeMsg.text?.match(/https:\/\/flowcrypt.com\/[a-z0-9A-Z]{10}/)) {
-      throw new HttpClientErr(`Error: cannot find pwd encrypted flowcrypt.com/api link in:\n\n${mimeMsg.text}`);
+      throw new HttpClientErr(`Error: cannot find pwd encrypted flowcrypt.com/shared-tenant-fes link in:\n\n${mimeMsg.text}`);
     }
```
Comment (tomholub) [inline test/source/mock/google/strategies/send-message-strategy.ts:63]: Same here

File: test/source/mock/lib/api.ts [hunk 21]
```diff
@@ -226,9 +226,10 @@ export class Api<REQ, RES> {
     if (body.length) {
       if (
         req.url!.startsWith('/upload/') || // gmail message send
-        req.url!.startsWith('/api/message/upload') || // flowcrypt.com/api pwd msg
+        req.url!.startsWith('/api/message/upload') || // flowcrypt.com/shared-tenant-fes pwd msg
```
Comment (tomholub) [inline test/source/mock/lib/api.ts:229]: Is there still any `/api/message/upload` endpoint?

File: test/source/tests/compose.ts [hunk 22]
```diff
@@ -2433,14 +2433,14 @@ export const defineComposeTests = (testVariant: TestVariant, testWithBrowser: Te
...
-        await composePage.checkElementColor('@input-password', 'rgb(49, 162, 23)');  // Password element color should turn into green (which means good password)
+        await composePage.checkElementColor('@input-password', 'rgb(49, 162, 23)'); // Password element color should turn into green (which means good password)
         await composePage.notPresent('#expiration_note'); // Expiration note should be hidden when password is good
         await composePage.waitAndClick('@action-send', { delay: 1 });
         await ComposePageRecipe.closed(composePage);
         const sentMsgs = (await GoogleData.withInitializedData(acct)).searchMessagesBySubject(subject);
         expect(sentMsgs.length).to.equal(2);
         // this test is using PwdEncryptedMessageWithFesIdTokenTestStrategy to check sent result based on subject "PWD encrypted message with FES - ID TOKEN"
-        // also see '/api/v1/message' in fes-endpoints.ts mock
+        // also see '/api/v1/message' in custom-url-fes-endpoints.ts mock
```
Comment (tomholub) [inline test/source/tests/compose.ts:2443]: `customer` + several below

File: test/source/tests/flaky.ts [hunk 23]
```diff
@@ -301,7 +301,7 @@ export const defineFlakyTests = (testVariant: TestVariant, testWithBrowser: Test
         expect((await GoogleData.withInitializedData(acct)).searchMessagesBySubject(subject).length).to.equal(expectedNumberOfPassedMessages); // + 0 messages
         // this test is using PwdEncryptedMessageWithFesReplyBadRequestTestStrategy to check sent result based on subject
         // "PWD encrypted message with FES web portal - some sends fail with BadRequest error"
-        // also see '/api/v1/message' in fes-endpoints.ts mock
+        // also see '/api/v1/message' in custom-url-fes-endpoints.ts mock
```
Comment (tomholub) [inline test/source/tests/flaky.ts:304]: customer

**Review by tomholub — COMMENTED** [review]
Implementation looks good, mostly just naming and details, listed below. 

Some tests are concerning / appear incorrect, and I haven't reviewed all tests yet.

File: extension/chrome/settings/index.ts [hunk 24]
```diff
@@ -438,7 +431,7 @@ View.run(
           Catch.reportErr(e);
           await Ui.modal.error(
             `There was an error changing google account, please ${Lang.general.contactMinimalSubsentence(
-              this.acctServer ? await this.acctServer.isFesUsed() : false
+              this.acctServer ? await isCustomerUrlFesUsed(this.acctEmail ?? '') : false
```
[the remark this answers is not in the document]
Comment (ioanmo226) [reply extension/chrome/settings/index.ts:434]: acctServer.isCustomerUrlFesUsed is removed as per your request.
https://github.com/FlowCrypt/flowcrypt-browser/pull/4907#discussion_r1087732236

[the remark this answers is not in the document]
Comment (tomholub) [reply extension/chrome/settings/index.ts:434]: I see. Let's leave it this way then.

File: extension/chrome/elements/compose-modules/compose-pwd-or-pubkey-container-module.ts (hunk 10 above)
Comment (ioanmo226) [reply extension/chrome/elements/compose-modules/compose-pwd-or-pubkey-container-module.ts:143]: ok

File: test/source/mock/backend/backend-data.ts (hunk 16 above)
Comment (ioanmo226) [reply test/source/mock/backend/backend-data.ts:41]: let me change it to clientConfigurationForDomain

File: test/source/mock/lib/api.ts (hunk 21 above)
Comment (ioanmo226) [reply test/source/mock/lib/api.ts:229]: I don't think so. let me remove it

File: test/source/mock/google/strategies/send-message-strategy.ts (hunk 19 above)
Comment (ioanmo226) [reply test/source/mock/google/strategies/send-message-strategy.ts:50]: @tomholub What will be exact url format?
I can't seem to get correct url format as live shared-tenant-fes is not available.

Comment (tomholub) [reply test/source/mock/google/strategies/send-message-strategy.ts]: > I can't seem to get correct url format as live shared-tenant-fes is not available.

It's available https://flowcrypt.com/shared-tenant-fes/

How are you attempting to test it?

Comment (ioanmo226) [reply test/source/mock/google/strategies/send-message-strategy.ts]: I tried with my gmail account and api says gmail.com client configuration is not available

Comment (ioanmo226) [reply test/source/mock/google/strategies/send-message-strategy.ts]: I'm asking about shared tenant password protected message url format though

Comment (tomholub) [reply test/source/mock/google/strategies/send-message-strategy.ts]: > I tried with my gmail account and api says gmail.com client configuration is not available

I see. I'll fix that. Then you can test with this, and you can then also exactly observe the url format.

In either case the url format will be exactly the same as when sending from flowcrypt.com domain, except that `fes.flowcrypt.com` becomes `flowcrypt.com/shared-tenant-fes`. The rest of the URL stays the same.

Comment (ioanmo226) [reply test/source/mock/google/strategies/send-message-strategy.ts]: Got it

File: extension/js/common/api/account-server.ts [hunk 25]
```diff
@@ -60,44 +46,24 @@ export class AccountServer extends Api {
...
+    const service = await this.getExternalService();
+    return await service.webPortalMessageNewReplyToken();
   };
 
-  private getIdToken = async (): Promise<string> => {
-    const idToken = await InMemoryStore.get(this.acctEmail, InMemoryStoreKeys.ID_TOKEN);
-    if (!idToken) {
-      // user will not actually see this message, they'll see a generic login prompt
-      throw new BackendAuthErr('Missing id token, please re-authenticate');
+  private getExternalService = async (): Promise<ExternalService> => {
```
Comment (sosnovsky) [inline extension/js/common/api/account-server.ts:63]: Maybe it's possible to check `isCustomerUrlFesUsed(this.acctEmail)` on `AccountServer` init and set service url there, so we don't need to run `getExternalService` before each request?

Comment (tomholub) [reply extension/js/common/api/account-server.ts:63]: I thought about that too. One advantage of current system could possibly be that it's a hair more dynamic - it will pick up on a change that happened after the class was initialized. But that's quite an edge case. I agree with what you say. I propose to do it in a separate PR.

File: test/source/mock/google/strategies/send-message-strategy.ts (hunk 19 above)
Comment (tomholub) [reply test/source/mock/google/strategies/send-message-strategy.ts]: It's configured now for gmail.com: https://flowcrypt.com/shared-tenant-fes/api/v1/client-configuration?domain=gmail.com

It's not configured yet for custom user domains, waiting for https://github.com/FlowCrypt/flowcrypt-enterprise-on-prem/issues/4162 to be released first

File: extension/js/common/api/account-server.ts (hunk 25 above)
Comment (sosnovsky) [reply extension/js/common/api/account-server.ts:63]: Yeah, can be a separate issue. Probably we can handle such edge case by re-initialising `AccountServer` when user's `fesUrl` property is changed.

Created issue https://github.com/FlowCrypt/flowcrypt-browser/issues/4917

Comment (tomholub) [reply extension/js/common/api/account-server.ts:63]: or we can just ignore the edge case, i think.

**Review by sosnovsky — APPROVED** [review]
Looks good for me 👍

**Comment by tomholub** [comment]
we'll still test it more after I finalize setup of shared-tenant-fes, before we release
